lib/streamer/cbor_streamer.go

```diff
@@ -0,0 +1,199 @@
+	file  *os.File
+	codec *codec.Encoder
+	wmu   sync.Mutex
+}
+
+type cborMuxRow struct {
+	Label int    `json:"l"`
+	Msg   []byte `json:"m,omitempty"`
+	Sig   int    `json:"x,omitempty"` // 1->closed
+}
+
+func CborFileMux(filePath string) Mux {
+	file, err := os.OpenFile(filePath, os.O_CREATE|os.O_RDWR|os.O_EXCL, 0755)
+	if err != nil {
+		panic(errors.IOError.Wrap(err))
+	}
+	file.Write([]byte{codec.CborStreamArray})
+	return &CborMux{
+		file:  file,
+		codec: codec.NewEncoder(file, new(codec.CborHandle)),
+	}
+	// consider using runtime.`SetFinalizer to close?  currently unhandled.
+}
+
+func (m *CborMux) write(label int, msg []byte) {
+	m.wmu.Lock()
+	defer m.wmu.Unlock()
+	const magic_RAW = 0
+	const magic_UTF8 = 1
+	//	m.codec.MustEncode(cborMuxRow{
+	//		Label: label,
+	//		Msg:   msg,
+	//	})
+	_, enc := codec.GenHelperEncoder(m.codec)
+	enc.EncodeMapStart(2)
+	enc.EncodeString(magic_UTF8, "l")
+	enc.EncodeInt(int64(label))
+	enc.EncodeString(magic_UTF8, "m")
+	enc.EncodeStringBytes(magic_RAW, msg)
+}
+
+func (m *CborMux) Close() {
+	m.wmu.Lock()
+	defer m.wmu.Unlock()
+	m.file.Write([]byte{0xff}) // should be `codec.CborStreamBreak`, plz update upstream for vis
+	// don't *actually* close, because readers can still be active on the same fd.
+	// TODO further writes should be forced into an error state
+}
+
+func (m *CborMux) Appender(label int) io.WriteCloser {
+	return &cborMuxAppender{m, label}
+}
+
+type cborMuxAppender struct {
+	m     *CborMux
+	label int
+}
+
+func (a *cborMuxAppender) Write(msg []byte) (int, error) {
+	a.m.write(a.label, msg)
+	return len(msg), nil
+}
+
+func (a *cborMuxAppender) Close() error {
+	a.m.wmu.Lock()
+	defer a.m.wmu.Unlock()
+	const magic_UTF8 = 1
+	_, enc := codec.GenHelperEncoder(a.m.codec)
+	enc.EncodeMapStart(2)
+	enc.EncodeString(magic_UTF8, "l")
+	enc.EncodeInt(int64(a.label))
+	enc.EncodeString(magic_UTF8, "x")
+	enc.EncodeInt(int64(1))
+	return nil
+}
+
+func (m *CborMux) Reader(labels ...int) io.Reader {
+	// asking for a reader for a label that was never used will never
+	// hit a close flag, so... don't do that?
+	r := io.NewSectionReader(m.file, 1, math.MaxInt64/2)
+	// TODO offset of one because that's the array open!
+	// do something much more sane than skip it, please
+	return &cborMuxReader{
+		labels: &intset{labels},
+		codec:  codec.NewDecoder(r, new(codec.CborHandle)),
+	}
+}
+
+type cborMuxReader struct {
+	labels *intset // remove them as we hit their close
+	codec  *codec.Decoder
+	buf    []byte // any remaining bytes from the last incomplete read
+}
+
+func (r *cborMuxReader) Read(msg []byte) (n int, err error) {
+	n, err = r.read(msg)
+	for n == 0 && err == nil {
+		// we're effectively required to block here, because otherwise the reader may spin;
+		// this is not a clueful wait; but it does prevent pegging a core.
+		// quite dumb in this case is also quite fool-proof.
+		time.Sleep(1 * time.Millisecond)
```

Comment:
Could I get your brief thoughts about upgrades to this in writing?
We discussed this in person but I think having a driveby of future ideas recorded would be wise.

I'm totally fine with landing bit, if that wasn't clear. Motivation is one of remembering